squad/frontend/tests.py

```diff
@@ -6,43 +8,99 @@
 from django.shortcuts import get_object_or_404
 
 
-class TestResult(object):
+class TestResult(list):
+    """
+    A List of pass/fail/skip statuses, one per environment. Represents one row
+    of the test results table.
+    """
 
     def __init__(self, name):
         self.name = name
-        self.results = []
 
-    def add(self, result):
-        self.results.append(result)
 
-    @property
-    def failure(self):
-        return any([r == 'fail' for r in self.results])
+class TestResultTable(list):
+    """
+    A plain list with a few extra attributes. Each list item represents one row
+    of the table, and should be an instance of TestResult.
 
-
-class TestResultTable(object):
+    This class also mimics a Django Paginator so that it can be used with our
+    pagination UI.
+    """
 
     def __init__(self):
         self.environments = None
-        self.failures = []
-        self.non_failures = []
-
-    def __bool__(self):
-        return bool(self.failures) or bool(self.non_failures)
-
-    def add(self, test):
-        if test.failure:
-            self.failures.append(test)
-        else:
-            self.non_failures.append(test)
+        self.paginator = self
+        self.paginator.num_pages = 0
+        self.number = 0
+
+    # count how many unique tests are represented in the given build, and sets
+    # pagination data
+    def __count_pages__(self, build_id, per_page):
+        query = """
+            SELECT COUNT(*)
+            FROM (
+                SELECT distinct suite_id, name
+                FROM core_test
+                JOIN core_testrun ON core_testrun.id = core_test.test_run_id
+                WHERE core_testrun.build_id = %s
+            ) unique_tests
+        """
+        with connection.cursor() as cursor:
+            cursor.execute(query, [build_id])
+            count = cursor.fetchone()[0]
+            # count pages
+            self.num_pages = count // per_page
+            if count % per_page > 0:
+                self.num_pages += 1
+
+    def __get_page_filter__(self, build_id, page, per_page):
+        """
+        Query to obtain one page os test results. It is used to know which tests
+        should be in the page. It's ordered so that the tests with more failures
+        come first, then tests with less failures, then tests with more skips.
```

Comment:
This feature doesn't work for me. I checked on data from production db " lkft » linux-mainline-oe » Build v4.15-rc7-111-g5f615b97cdea". The failed tests don't show up on 1st or second page. Skips are on a few last pages.

Reply:
weird, it works for me here:

![screen shot 2018-01-23 at 17 11 32](https://user-images.githubusercontent.com/30027/35295516-80b9d488-0060-11e8-847d-431e1a031aab.png)

Reply:
Antonio, this is a different view. I thought the PR is about 'all test results'. You're showing the paginated result for single TestRun. That works for me as well. 

Here is what I'm getting in 'all test results':
![all_test_results](https://user-images.githubusercontent.com/618901/35325396-8d4dda9c-00eb-11e8-9a18-f2ff05d09842.png)

Reply:
yes! :-(

I found the issue, and pushed updates to this PR